src/server.ts

```diff
@@ -81,66 +83,178 @@ export async function buildServer (ctx: ViteBuildContext) {
-  await writeFile(resolve(serverDist, 'client.manifest.json'), JSON.stringify({
-    publicPath: '',
-    all: [],
+  const SPA_TEMPLATE = APP_TEMPLATE
+    .replace('{{ APP }}', '<div id="__nuxt">{{ APP }}</div>')
+    .replace(
+      '</body>',
+      '<script type="module" src="/@vite/client"></script><script type="module" src="/client.js"></script></body>'
+    )
+  const SSR_TEMPLATE = ctx.nuxt.options.dev ? SPA_TEMPLATE : APP_TEMPLATE
+
+  await writeFile(r('index.ssr.html'), SSR_TEMPLATE)
+  await writeFile(r('index.spa.html'), SPA_TEMPLATE)
+
+  if (ctx.nuxt.options.dev) {
+    await stubManifest(ctx)
+  } else {
+    await generateBuildManifest(ctx)
+  }
+
+  const onBuild = () => ctx.nuxt.callHook('build:resources', wpfs)
+  const build = async () => {
+    const start = Date.now()
+    await vite.build(serverConfig)
+    await onBuild()
+    consola.info(`Server built in ${Date.now() - start}ms`)
+  }
+
+  const debouncedBuild = debounce(build, 300)
+
+  await build()
+
+  if (ctx.nuxt.options.dev) {
+    const watcher = watch([
+      ctx.nuxt.options.buildDir,
+      ctx.nuxt.options.srcDir,
+      ctx.nuxt.options.rootDir
+    ], {
+      ignored: [
+        '**/dist/server/**'
+      ]
+    })
+
+    watcher.on('change', () => debouncedBuild())
+
+    ctx.nuxt.hook('close', async () => {
+      await watcher.close()
+    })
+  }
+}
+
+// convert vite's manifest to webpack style
+async function generateBuildManifest (ctx: ViteBuildContext) {
+  const serverDist = resolve(ctx.nuxt.options.buildDir, 'dist/server')
+  const clientDist = resolve(ctx.nuxt.options.buildDir, 'dist/client')
+  const r = (...args: string[]): string => resolve(serverDist, ...args)
+
+  const publicPath = ctx.nuxt.options.app.assetsPath // Default: /nuxt/
+  const viteClientManifest = await readJSON(join(clientDist, 'manifest.json'))
+
+  function getModuleIds ([, value]: [string, any]) {
+    if (!value) {
+      return []
+    }
+    return [value.file, ...value.css || []]
+      .filter(i => !i.endsWith('.js') || i.match(/-legacy\./)) // only use legacy build
+  }
+
+  const asyncEntries = uniq(
+    Object.entries(viteClientManifest)
+      .filter((i: any) => i[1].isDynamicEntry)
+      .flatMap(getModuleIds)
+  ).filter(i => i)
+  const initialEntries = uniq(
+    Object.entries(viteClientManifest)
+      .filter((i: any) => !i[1].isDynamicEntry)
+      .flatMap(getModuleIds)
+  ).filter(i => i)
+  const initialJs = initialEntries.filter(i => i.endsWith('.js'))
+  const initialAssets = initialEntries.filter(i => !i.endsWith('.js'))
+
+  // search for polyfill file, we don't include it in the client entry
+  const polyfillName = initialEntries.find(i => i.startsWith('polyfills-legacy.'))
+
+  // @vitejs/plugin-legacy uses SystemJS which need to call `System.import` to load modules
+  const clientEntryCode = initialJs
+    .filter(i => !i.startsWith('polyfills-legacy.'))
+    .map(i => `System.import("${publicPath}${i}");`)
```

Comment:
Don't we need to `import` sequentially? Currently vendor is imported after client in entry like this: 

```js
System.import("/_nuxt/client-legacy.ccba9bb0.js");
System.import("/_nuxt/vendor-legacy.df48698c.js");
```

(notice that Promises are discarded)

Reply:
Yeah good point, agree we should wrap with an iife and do await sequentially

I see vite is injecting this without handling the error:
https://github.com/vitejs/vite/blob/024a2de63df60a4037f9f7b13a0bc6e2b0d41fb6/packages/plugin-legacy/index.js

Reply:
https://github.com/nuxt/vite/pull/179/commits/ed52d34876ef74fcd97d5bedd8a1e261335bca23